cpp/src/arrow/scalar.cc

```diff
@@ -1127,6 +1127,28 @@ Status CastImpl(const StructScalar& from, StringScalar* to) {
   return Status::OK();
 }
 
+// casts between variable-length and fixed-length list types
+template <typename ToScalar>
+enable_if_t<is_list_like_type<typename ToScalar::TypeClass>::value &&
+                ToScalar::TypeClass::type_id != Type::MAP,
+            Status>
+CastImpl(const BaseListScalar& from, ToScalar* to) {
+  DCHECK(from.type->id() == Type::LIST || from.type->id() == Type::LARGE_LIST ||
+         from.type->id() == Type::FIXED_SIZE_LIST);
```

Comment:
Careful: `MapScalar` inherits from `BaseListScalar`. Are you sure it can't use this overload?

Reply:
I will use `is_var_length_list` in the enable-if and expand the test cases.

Reply:
Used a better predicate for the enable_if here and extended it to gracefully handle Map as well.